symbolic-unwind/src/evaluator/mod.rs

```diff
@@ -0,0 +1,474 @@
+//! Functionality for evaluating *Breakpad
+//! [RPN](https://en.wikipedia.org/wiki/Reverse_Polish_notation) expressions*.
+//!
+//! These expressions are defined by the following
+//! [BNF](https://en.wikipedia.org/wiki/Backus%E2%80%93Naur_form) specification:
+//! ```text
+//! <expr>     ::=  <contant> | <variable> | <literal> | <expr> <expr> <binop> | <expr> ^
+//! <constant> ::=  [a-zA-Z_.][a-zA-Z0-9_.]*
+//! <variable> ::=  $[a-zA-Z][a-zA-Z0-9]*
+//! <binop>    ::=  + | - | * | / | % | @
+//! <literal>  ::=  [0-9a-fA-F]+
+//! ```
+//! Most of this syntax should be familiar. The symbol `^` denotes a dereference operation,
+//! i.e. assuming that some representation `m` of a region of memory is available,
+//! `x ^` evaluates to `m[x]`. If no memory is available or `m` is not defined at `x`, the
+//! expression's value is undefined. The symbol
+//! `@` denotes an align operation; it truncates its first operand to a multiple of its
+//! second operand.
+//!
+//! Constants and variables are evaluated by referring to dictionaries
+//! (concretely: [`BTreeMap`]s). If an expression contains a constant or variable that is
+//! not in the respective dictionary, the expression's value is undefined.
+//!
+//! In addition to expressions, there are also *assignments*:
+//! ```text
+//! <assignment> ::=  <variable> <expr> =
+//! ```
+//! An assignment results in an update of the variable's value in the dictionary, or its
+//! insertion if it was not defined before.
+use std::collections::{BTreeMap, BTreeSet};
+use std::error::Error;
+use std::fmt;
+use std::str::FromStr;
+
+use super::base::{Endianness, MemoryRegion, RegisterValue};
+use parsing::ParseExprError;
+
+pub mod parsing;
+/// Structure that encapsulates the information necessary to evaluate Breakpad
+/// RPN expressions.
+///
+/// It is generic over:
+/// - An address type, which is used both for basic expressions and for pointers into `memory`
+/// - An [`Endianness`](super::base::Endianness) that controls how values are read from memory
+pub struct Evaluator<'memory, A, E> {
+    /// A region of memory.
+    ///
+    /// If this is `None`, evaluation of expressions containing dereference
+    /// operations will fail.
+    memory: Option<MemoryRegion<'memory>>,
+
+    /// A map containing the values of constants.
+    ///
+    /// Trying to use a constant that is not in this map will cause evaluation to fail.
+    constants: BTreeMap<Constant, A>,
+
+    /// A map containing the values of variables.
+    ///
+    /// Trying to use a variable that is not in this map will cause evaluation to fail.
```

Comment:
This contradicts the module level docs, since they mention "undefined". I do prefer returning a failure over some kind of UB, so we should clear up the docs above.

Reply:
Ah, I meant "the value of the expression is undefined" in the sense of "evaluating the expression will fail", not in the sense of UB. Definitely a good idea to clarify this.